chainntnfs/btcdnotify/btcd.go

```diff
@@ -290,80 +276,68 @@ out:
-			// update from the front of the update queue.
-			b.chainUpdateMtx.Lock()
-			update := b.chainUpdates[0]
-			b.chainUpdates[0] = nil // Set to nil to prevent GC leak.
-			b.chainUpdates = b.chainUpdates[1:]
-			b.chainUpdateMtx.Unlock()
-
-			currentHeight = update.blockHeight
-
-			newBlock, err := b.chainConn.GetBlock(update.blockHash)
-			if err != nil {
-				chainntnfs.Log.Errorf("Unable to get block: %v", err)
-				continue
-			}
+		case item := <-b.chainUpdates.ChanOut():
+			update := item.(*chainUpdate)
+			if update.connect {
+				if update.blockHeight != currentHeight+1 {
+					chainntnfs.Log.Warnf("Received blocks out of order: "+
+						"current height=%d, new height=%d",
+						currentHeight, update.blockHeight)
+				}
 
-			chainntnfs.Log.Infof("New block: height=%v, sha=%v",
-				update.blockHeight, update.blockHash)
-
-			b.notifyBlockEpochs(update.blockHeight,
-				update.blockHash)
-
-			newHeight := update.blockHeight
-			for i, tx := range newBlock.Transactions {
-				// Check if the inclusion of this transaction
-				// within a block by itself triggers a block
-				// confirmation threshold, if so send a
-				// notification. Otherwise, place the
-				// notification on a heap to be triggered in
-				// the future once additional confirmations are
-				// attained.
-				txSha := tx.TxHash()
-				b.checkConfirmationTrigger(&txSha, update, i)
-			}
+				currentHeight = update.blockHeight
+
+				rawBlock, err := b.chainConn.GetBlock(update.blockHash)
+				if err != nil {
+					chainntnfs.Log.Errorf("Unable to get block: %v", err)
+					continue
+				}
+
+				chainntnfs.Log.Infof("New block: height=%v, sha=%v",
+					update.blockHeight, update.blockHash)
+
+				b.notifyBlockEpochs(update.blockHeight, update.blockHash)
 
-			// A new block has been connected to the main
-			// chain. Send out any N confirmation notifications
-			// which may have been triggered by this new block.
-			b.notifyConfs(newHeight)
+				txns := btcutil.NewBlock(rawBlock).Transactions()
+				err = b.txConfNotifier.ConnectTip(update.blockHash,
+					uint32(update.blockHeight), txns)
+				if err != nil {
+					chainntnfs.Log.Error(err)
+				}
+			} else {
+				if update.blockHeight != currentHeight {
+					chainntnfs.Log.Warnf("Received blocks out of order: "+
+						"current height=%d, disconnected height=%d",
+						currentHeight, update.blockHeight)
+				}
 
-		case <-b.txUpdateSignal:
-			// A new update is available, so pop the new chain
-			// update from the front of the update queue.
-			b.txUpdateMtx.Lock()
-			newSpend := b.txUpdates[0]
-			b.txUpdates[0] = nil // Set to nil to prevent GC leak.
-			b.txUpdates = b.txUpdates[1:]
-			b.txUpdateMtx.Unlock()
+				currentHeight = update.blockHeight - 1
 
+				chainntnfs.Log.Infof("Block disconnected from main chain: "+
+					"height=%v, sha=%v", update.blockHeight, update.blockHash)
+
+				err := b.txConfNotifier.DisconnectTip(uint32(update.blockHeight))
```

Comment:
Maybe a dumb question, but isn't a reorg possible at a block that's not the latest? Should this call `DisconnectTip` at each block height between the current `currentHeight` and `update.blockHeight` (inclusive)?

Reply:
`btcd` and `neutrino` reorgs both fire a `disconnectblock` notification for each block disconnected in a reorg. Makes it more convenient to work with than `bitcoind`'s notifications, which only notify you of the latest block in such a situation.